### PR TITLE
Refactor sqlalchemy queries to 2.0 style (Part 1) 

### PR DESCRIPTION
**Pull request by ephraimbuddy**
This commit updates the sqlalchemy queries to adopt the 2.0 query style,
which is compatible with version 1.4. The changes involve updating the engine
with the future=True flag to indicate the execution of queries using the 2.0 style.
As a result, textual SQL statements are wrapped with the text function.

In addition, queries that previously used delete and update operations have been
modified to utilize the new delete/update construct. Furthermore, all queries within the
jobs/ and api/ directories have been thoroughly updated to employ the new style queries.

Please note that this commit intentionally stops at this point to ensure ease of review for the pull request.
The only test change is the addition of the future flag to the create_engine function.

### CODE REVIEW

File: airflow/api/common/mark_tasks.py [hunk 1]
```diff
@@ -160,10 +160,10 @@ def set_state(
             task_instance.set_state(state, session=session)
         session.flush()
     else:
-        tis_altered = qry_dag.all()
+        tis_altered = session.execute(qry_dag).scalars().all()
         if sub_dag_run_ids:
             qry_sub_dag = all_subdag_tasks_query(sub_dag_run_ids, session, state, confirmed_dates)
-            tis_altered += qry_sub_dag.all()
+            tis_altered += session.execute(qry_sub_dag.scalars().all())
```
Comment (uranusjr) [inline airflow/api/common/mark_tasks.py]:
```suggestion
            tis_altered += session.execute(qry_sub_dag).scalars().all()
```

Wrong paranthesis?

File: airflow/api/common/mark_tasks.py [hunk 2]
```diff
@@ -464,13 +463,16 @@ def set_dag_run_state_to_failed(
...
+        select(TaskInstance).where(
+            TaskInstance.dag_id == dag.dag_id,
+            TaskInstance.run_id == run_id,
+            TaskInstance.task_id.in_(task_ids),
+            TaskInstance.state.in_([State.RUNNING, State.DEFERRED, State.UP_FOR_RESCHEDULE]),
+        )
     )
-    task_ids_of_running_tis = [task_instance.task_id for task_instance in tis]
+
+    task_ids_of_running_tis = [task_instance.task_id for task_instance, in tis]
```
Comment (uranusjr) [inline airflow/api/common/mark_tasks.py:475]: We probably can rewrite this with `session.scalars` on `TaskInstance.task_id`?

Comment (uranusjr) [inline airflow/api/common/mark_tasks.py]: Note: This can be rewritten with `scalars` on `TaskInstance.task_id` instead. (Maybe in a later PR.)

File: airflow/api/common/mark_tasks.py [hunk 3]
```diff
@@ -480,14 +482,16 @@ def set_dag_run_state_to_failed(
...
+        select(TaskInstance).filter(
+            TaskInstance.dag_id == dag.dag_id,
+            TaskInstance.run_id == run_id,
+            TaskInstance.state.not_in(State.finished),
+            TaskInstance.state.not_in([State.RUNNING, State.DEFERRED, State.UP_FOR_RESCHEDULE]),
+        )
     )
 
-    tis = [ti for ti in tis]
+    tis = [ti for ti, in tis]
```
Comment (uranusjr) [inline airflow/api/common/mark_tasks.py]: Use `scalars` instead?

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 4]
```diff
@@ -74,7 +74,10 @@
 @provide_session
 def delete_dag_run(*, dag_id: str, dag_run_id: str, session: Session = NEW_SESSION) -> APIResponse:
     """Delete a DAG Run."""
-    if session.query(DagRun).filter(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id).delete() == 0:
+    if (
+        session.execute(delete(DagRun).where(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id)).rowcount
+        == 0
+    ):
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
    deleted_count = session.execute(delete(DagRun).where(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id)).rowcount
    if deleted_count == 0:
```

File: airflow/jobs/backfill_job_runner.py [hunk 5]
```diff
@@ -183,7 +184,7 @@ def _update_counters(self, ti_status: _DagRunTaskStatus, session: Session) -> No
 
         filter_for_tis = TI.filter_for_tis(list(ti_status.running.values()))
         if filter_for_tis is not None:
-            refreshed_tis = session.query(TI).filter(filter_for_tis).all()
+            refreshed_tis = session.execute(select(TI).where(filter_for_tis)).scalars().all()
```
Comment (uranusjr) [inline airflow/jobs/backfill_job_runner.py]:
```suggestion
            refreshed_tis = session.scalars(select(TI).where(filter_for_tis)).all()
```

Is this the same? (p.s. I think the `all()` is not needed either but that should probably be another PR)

File: airflow/jobs/job.py [hunk 6]
```diff
@@ -252,16 +252,16 @@ def most_recent_job(job_type: str, session: Session = NEW_SESSION) -> Job | None
     :param job_type: job type to query for to get the most recent job for
     :param session: Database session
     """
-    return (
-        session.query(Job)
-        .filter(Job.job_type == job_type)
+    return session.scalars(
```
Comment (uranusjr) [inline airflow/jobs/job.py]: Can this use `session.scalar(...)` instead to eliminate the `first()` call?

File: airflow/jobs/scheduler_job_runner.py [hunk 7]
```diff
@@ -393,7 +394,8 @@ def _executable_task_instances_to_queued(self, max_tis: int, session: Session) -
                     of=TI,
                     session=session,
                     **skip_locked(session=session),
-                ).all()
+                )
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py:397]: This makes the variable’s annotation wrong

File: airflow/jobs/scheduler_job_runner.py [hunk 8]
```diff
@@ -1199,9 +1210,11 @@ def _create_dag_runs_dataset_triggered(
...
         }
         existing_dagruns: set[tuple[str, timezone.DateTime]] = set(
-            session.query(DagRun.dag_id, DagRun.execution_date).filter(
-                tuple_in_condition((DagRun.dag_id, DagRun.execution_date), exec_dates.items())
-            )
+            session.execute(
+                select(DagRun.dag_id, DagRun.execution_date).where(
+                    tuple_in_condition((DagRun.dag_id, DagRun.execution_date), exec_dates.items())
+                )
+            ).unique()
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py]: Arguably not needed since this is going into a set, but not harm anyway.

File: airflow/jobs/backfill_job_runner.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/jobs/backfill_job_runner.py]: Yes. Both are the same. I agree to a different PR to remove` all()`

File: airflow/jobs/job.py (hunk 6 above)
Comment (ephraimbuddy) [reply airflow/jobs/job.py]: I wonder why they didn't use scalar in their 1.4/2 examples? scalar worked. Will need to do the same for other places but not sure why they didn't use it as an example

Comment (ephraimbuddy) [reply airflow/jobs/job.py]: I'm not sure we have tests for when the item is more than 1. scalar may order these items randomly and pick the first. WDYT?

File: airflow/jobs/scheduler_job_runner.py (hunk 7 above)
Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job_runner.py:397]: Fixed

File: airflow/jobs/job.py (hunk 6 above)
Comment (uranusjr) [reply airflow/jobs/job.py]: > scalar may order these items randomly and pick the first

Yeah, but wouldn’t `scalars` and then `first` be subject to the same issue? If I understand things correctly the two are essentially equivalent.

File: airflow/api/common/mark_tasks.py [hunk 9]
```diff
@@ -148,10 +148,10 @@ def set_state(
     qry_dag = get_all_dag_task_query(dag, session, state, task_id_map_index_list, dag_run_ids)
 
     if commit:
-        tis_altered = qry_dag.with_for_update().all()
+        tis_altered = session.execute(qry_dag.with_for_update()).scalars().all()
         if sub_dag_run_ids:
             qry_sub_dag = all_subdag_tasks_query(sub_dag_run_ids, session, state, confirmed_dates)
-            tis_altered += qry_sub_dag.with_for_update().all()
+            tis_altered += session.execute(qry_sub_dag.with_for_update()).scalars().all()
```
Comment (uranusjr) [inline airflow/api/common/mark_tasks.py]:
```suggestion
            tis_altered += session.scalars(qry_sub_dag.with_for_update()).all()
```

I’m not quite sure but the docs seem to suggest this is superior…?

https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#selecting-orm-entities

> This is most easily achieved by using the Session.scalars() method to execute, rather than the Session.execute()

File: images/breeze/output_ci.svg
Comment (uranusjr) [inline images/breeze/output_ci.svg:1]: Hmm, it’s weird these screenshots are being updated. The added colouring also is suspective.

File: airflow/jobs/job.py (hunk 6 above)
Comment (ephraimbuddy) [reply airflow/jobs/job.py]: I have confirmed why they didn't use it, it's slow on big data. `session.scalar(select(TaskInstance))` got stuck on 17m rows taskinstance

Comment (uranusjr) [reply airflow/jobs/job.py]: Adding a `limit(1)` would fix it, I think?

File: airflow/api/common/mark_tasks.py (hunk 9 above)
Comment (ephraimbuddy) [reply airflow/api/common/mark_tasks.py]: Used session.scalars where possible https://github.com/apache/airflow/pull/31569/commits/37edb980f31fbe13e0de14ede679519cb5ec90fc

File: images/breeze/output_ci.svg
Comment (ephraimbuddy) [reply images/breeze/output_ci.svg:1]: I ran `breeze setup regenerate-command-images --force` when I couldn't find why the PR was failing on pre-commit generating ER diagram(not locally)

Comment (uranusjr) [reply images/breeze/output_ci.svg:1]: Yes, what I was trying to say is I don’t understand why pre-commit thinks they need to be updated at all.

File: airflow/jobs/job.py [hunk 10]
```diff
@@ -252,16 +252,16 @@ def most_recent_job(job_type: str, session: Session = NEW_SESSION) -> Job | None
...
+        .where(Job.job_type == job_type)
         .order_by(
             # Put "running" jobs at the front.
             case({State.RUNNING: 0}, value=Job.state, else_=1),
             Job.latest_heartbeat.desc(),
         )
-        .first()
-    )
+        .limit(1)
+    ).first()
```
Comment (uranusjr) [inline airflow/jobs/job.py]:
```python
return session.scalar(
    select(Job)
    .where(Job.job_type == job_type)
    .order_by(...)
    .limit(1)
)
```

I think this would be equivalent?

File: airflow/api/common/mark_tasks.py [hunk 11]
```diff
@@ -361,7 +358,9 @@ def _set_dag_run_state(dag_id: str, run_id: str, state: DagRunState, session: SA
     :param state: target state
     :param session: database session
     """
-    dag_run = session.query(DagRun).filter(DagRun.dag_id == dag_id, DagRun.run_id == run_id).one()
+    dag_run = session.execute(
+        select(DagRun).where(DagRun.dag_id == dag_id, DagRun.run_id == run_id)
+    ).scalar_one()
```
Comment (uranusjr) [inline airflow/api/common/mark_tasks.py:363]: Should be equivalent to

```suggestion
    dag_run = session.scalars(
        select(DagRun).where(DagRun.dag_id == dag_id, DagRun.run_id == run_id)
    ).one()
```

but I _don’t_ think this is better. So just for reference in case anyone wonders.

**Review by uranusjr — APPROVED** [review]
